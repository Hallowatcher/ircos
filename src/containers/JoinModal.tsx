
import * as React from 'react';
import Modal from 'react-modal';

interface IState {
  isOpen: boolean;
  onClose: () => void;
  joinChannel: (event: any) => void;
}

const styles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)'
  }
};

export class JoinModal extends React.Component<IState, any> {
  constructor() {
    super();

    this.state = {
      channel: ''
    };

    this.joinChannel = this.joinChannel.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  public render() {
    return (
      <Modal
        isOpen={this.props.isOpen}
        onRequestClose={this.props.onClose}
        contentLabel='Join channel'
        style={styles}
      >
        <h1>Join channel modal</h1>
        <form onSubmit={this.joinChannel}>
          <input placeholder="#channel or user" onChange={this.handleChange} />
          <input type="submit" />
        </form>
      </Modal>
    );
  }

  private handleChange(event: any) {
    this.setState({ channel: event.target.value });
  }

  private joinChannel(event: any) {
    event.preventDefault();

    if (this.state.channel !== '') {
      this.props.joinChannel(this.state.channel);
      this.props.onClose();
    }
  }
}
