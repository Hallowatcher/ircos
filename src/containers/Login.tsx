import * as React from 'react';
import { connect } from 'react-redux';
import { LoginForm } from '../components/LoginForm';
import { createConnection } from '../actions/client';

type Props = {
  createConnection: any
}

const divStyle: React.CSSProperties = {
  margin: 0,
  padding: 0,
  width: '100%',
  height: '100%',
  backgroundColor: '#acadaf',
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  alignContent: 'center'
}
const divChild: React.CSSProperties = {
  backgroundColor: 'white',
  alignSelf: 'center',
  boxShadow: '0 0 4px rgba(0, 0, 0, 0.35)'
}

export class Login extends React.Component<Props, null> {
  render() {
    return (
      <div style={divStyle}>
        <div style={divChild}>
          <LoginForm submitLogin={this.props.createConnection.bind(this)} />
        </div>
      </div>
    )
  }
}

/* istanbul ignore next */
function dispatchToProps(dispatch: any) {
   return {
    createConnection: (user: string, pass: string) => { dispatch( createConnection(user, pass) ) }
  }
}

/* istanbul ignore next */
export default connect<any, any, any>(null, dispatchToProps)(Login);