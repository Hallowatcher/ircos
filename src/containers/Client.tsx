import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { connect } from 'react-redux';

import {
  makeCurrentChannel,
  sendMessage,
  leaveChannel,
  join
} from '../actions/client';

import { Tab } from '../components/Tab';
import { AddTab } from '../components/AddTab';
import { ChatView } from '../components/ChatView';
import { JoinModal } from './JoinModal';

type StateProps = {
  channels: any,
  messages: any[],
  currentChannel: string,
  nick: string,
  userId: number,
  channelLength: number
}

type DispatchProps = {
  makeCurrentChannel: (channel: string) => void,
  sendMessage: (channel: string, message: string) => void,
  closeChannel: (channel: string) => void ,
  joinChannel: (channel: string) => void
}

let styles = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
    height: '100%',
    overflow: 'hidden'
  } as React.CSSProperties,
  topBar: {
    backgroundColor: '#FE4590',
    flexShrink: 0,
    display: 'flex',
    flexDirection: 'row'
  } as React.CSSProperties,
  tabs: {
    flexGrow: 1,
    padding: '0 5px',
    display: 'flex',
    overflowY: 'hidden',
    overflowX: 'auto',
    alignItems: 'flex-end'
  } as React.CSSProperties,
  currentUser: {
    backgroundColor: 'rgba(255, 255, 255, 0.4)',
    display: 'flex',
    color: 'white'
  } as React.CSSProperties,
  currentUserImageContainer: {
    height: '35px',
    width: '35px',
    backgroundColor: 'rgba(255, 255, 255, 0.8)'
  } as React.CSSProperties,
  currentUserImage: {
    maxHeight: '100%'
  } as React.CSSProperties,
  currentUserNick: {
    padding: 5,
    display: 'flex',
    alignItems: 'center'
  } as React.CSSProperties,
  chat: {
    flexGrow: 1,
    overflowY: 'auto',
    overflowX: 'hidden',
    wordBreak: 'break-word'
  } as React.CSSProperties,
  dummyDiv: {
    float: 'left',
    clear: 'both'
  } as React.CSSProperties
}

export class Client extends React.Component<StateProps & DispatchProps, any> {

  constructor(props: any) {
    super(props);

    this.state = {
      msg: '',
      showJoinModal: false
    }
  }

  displayMessages = [];
  displayMessagesAmount: number = 50

  handleChangeMsg(event: any) {
    this.setState({ msg: event.target.value });
  }

  handleSubmit(event: any) {
    event.preventDefault();

    this.props.sendMessage(this.props.currentChannel, this.state.msg);
    this.setState({ msg: '' });
  }

  render() {

    // Tabs
    let channels = Object.keys(this.props.channels);
    let channelMap = channels.map((channel, index) => 
      <Tab
        key={index}
        tabName={channel}
        tabClick={this.props.makeCurrentChannel.bind(this)}
        closeTab={this.props.closeChannel.bind(this)}
        isActive={this.props.currentChannel === channel}
      />
    );

    let userImage = (() => {
      if (this.props.userId !== 0)
        return <img style={styles.currentUserImage} src={`https://a.ppy.sh/${this.props.userId}_${Date.now()}.jpg`} />;
    })();

    return (
      <div style={styles.container}>

        {/*Modals*/}
        <JoinModal
          isOpen={this.state.showJoinModal}
          onClose={() => { this.setState({ showJoinModal: false }) }}
          joinChannel={ this.props.joinChannel.bind(this) }
        />

        {/*Top bar*/}
        <div style={styles.topBar}>

          {/*Tabs*/}
          <div style={styles.tabs}>
            {channelMap}
            <AddTab clickAddTab={() => this.setState({ showJoinModal: true })} />
          </div>

          {/*Current user*/}
          <div style={styles.currentUser}>
            <div style={styles.currentUserImageContainer}>{userImage}</div>
            <div style={styles.currentUserNick}>{this.props.nick}</div>
          </div>

        </div>

        {/*Chat container*/}
        <ChatView
          style={styles.chat}
          messages={this.props.messages}
          selfNick={this.props.nick}
        />

        {/*Send message*/}
        <form onSubmit={this.handleSubmit.bind(this)}>
          <input type="text" placeholder="Enter a message..." value={this.state.msg} onChange={this.handleChangeMsg.bind(this)} />
          <input type="submit" value="Submit" />
        </form>
        
      </div>
    )
  }
}

/* istanbul ignore next */
function stateToProps(state: any) {
  return {
    channels: state.get('channelDb').toJS(),
    messages: state.getIn(['channelCurrent', 'messages']).toJS(),
    currentChannel: state.getIn(['channelCurrent', 'name']),
    nick: state.getIn(['userInfo', 'userName']),
    userId: state.getIn(['userInfo', 'userId']),
    channelLength: state.getIn(['settings', 'channelLength'])    
  }
}

/* istanbul ignore next */
function dispatchToProps(dispatch: any) {
  return {
    makeCurrentChannel: (channel: string) => { dispatch( makeCurrentChannel(channel) ) },
    sendMessage: (channel: string, message: string) => { dispatch( sendMessage(channel, message) ) },
    closeChannel: (channel: string) => { dispatch( leaveChannel(channel) ) },
    joinChannel: (channel: string) => { dispatch( join(channel) ) }
  }
}

/* istanbul ignore next */
export default connect(stateToProps, dispatchToProps)(Client);