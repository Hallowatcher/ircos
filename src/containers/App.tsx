import * as React from 'react';
import { connect } from 'react-redux';

// Drag n Drop
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';

import Login from './Login';
import Client from './Client';

interface IStateProps {
  connected: boolean;
}

export class App extends React.Component<IStateProps, any> {

  public render() {
    if (!this.props.connected) {
      return <Login />;
    }

    return <Client />;
  }
}

/* istanbul ignore next */
function stateToProps(state: any): IStateProps {
  return {
    connected: !!state.getIn(['serverInfo', 'connected'])
  };
}

/* istanbul ignore next */
export default connect<IStateProps, any, any>(stateToProps)(DragDropContext(HTML5Backend)(App));
