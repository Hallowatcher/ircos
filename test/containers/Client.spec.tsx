import * as React from 'react'
import { Client } from '../../src/containers/Client'
import * as enzyme from 'enzyme'
import * as MockDate from 'mockdate'
import toJson from 'enzyme-to-json'

describe('Client', function () {

  let component: enzyme.ShallowWrapper<any, any>
  let datenow
  let sendMessage = jest.fn()
  beforeEach(function () {
    datenow = 1262304000000;

    // Mock date
    MockDate.set(datenow)

    component = enzyme.shallow(
      <Client 
        nick={''}
        channelLength={5}
        channels={[]}
        closeChannel={jest.fn()}
        currentChannel={''}
        joinChannel={jest.fn()}
        makeCurrentChannel={jest.fn()}
        sendMessage={sendMessage}
        userId={1}
        messages={[]}
      />
    )
  })

  afterEach(() => {
    MockDate.reset()
  })

  it('should render', function () {
    expect(component).toMatchSnapshot()
  })

  it('should mount', function () {
    let mountedComponent = enzyme.mount(
      <Client 
        nick={''}
        channelLength={5}
        channels={[]}
        closeChannel={jest.fn()}
        currentChannel={''}
        joinChannel={jest.fn()}
        makeCurrentChannel={jest.fn()}
        sendMessage={jest.fn()}
        userId={1}
        messages={[]}
      />
    )
  })

  it('should mount with userid = 0', function () {
    let mountedComponent = enzyme.mount(
      <Client 
        nick={''}
        channelLength={5}
        channels={[]}
        closeChannel={jest.fn()}
        currentChannel={''}
        joinChannel={jest.fn()}
        makeCurrentChannel={jest.fn()}
        sendMessage={jest.fn()}
        userId={0}
        messages={[]}
      />
    )
  })

  it('should mount with channel tabs', function () {
    let mountedComponent = enzyme.mount(
      <Client 
        nick={''}
        channelLength={5}
        channels={['one', 'two']}
        closeChannel={jest.fn()}
        currentChannel={''}
        joinChannel={jest.fn()}
        makeCurrentChannel={jest.fn()}
        sendMessage={jest.fn()}
        userId={1}
        messages={[]}
      />
    )

    expect(toJson(mountedComponent)).toMatchSnapshot();
  })

  it('should open join modal', function () {
    component.find('AddTab').first().props()['clickAddTab']();
    expect(component.state('showJoinModal')).toEqual(true);
  })

  it('should close join modal', function () {
    component.find('JoinModal').first().props()['onClose']();
    expect(component.state('showJoinModal')).toEqual(false);
  })

  it('should change message', function () {
    component.find('input[type="text"]').first()
      .simulate('change', { target: { value: 'changed' } })
    expect(component.state('msg')).toEqual('changed')
  })

  it('should send message', function () {
    component.find('input[type="text"]').first()
      .simulate('change', { target: { value: 'changed' } })
    component.find('form').first()
      .simulate('submit', { preventDefault() {} })
    expect(sendMessage).toHaveBeenCalledWith('', 'changed')
    expect(component.state('msg')).toEqual('')
  })

})