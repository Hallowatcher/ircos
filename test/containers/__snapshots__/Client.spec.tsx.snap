// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client should mount with channel tabs 1`] = `
<Client
  channelLength={5}
  channels={
    Array [
      "one",
      "two",
    ]
  }
  closeChannel={[Function]}
  currentChannel=""
  joinChannel={[Function]}
  makeCurrentChannel={[Function]}
  messages={Array []}
  nick=""
  openExternal={[Function]}
  sendMessage={[Function]}
  userId={1}
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "overflow": "hidden",
        "width": "100%",
      }
    }
  >
    <JoinModal
      isOpen={false}
      joinChannel={[Function]}
      onClose={[Function]}
    >
      <Modal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        contentLabel="Join channel"
        isOpen={false}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="ReactModalPortal"
        shouldCloseOnOverlayClick={true}
        style={
          Object {
            "content": Object {
              "bottom": "auto",
              "left": "50%",
              "marginRight": "-50%",
              "right": "auto",
              "top": "50%",
              "transform": "translate(-50%, -50%)",
            },
          }
        }
      />
    </JoinModal>
    <div
      style={
        Object {
          "backgroundColor": "#FE4590",
          "display": "flex",
          "flexDirection": "row",
          "flexShrink": 0,
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "flex-end",
            "display": "flex",
            "flexGrow": 1,
            "overflowX": "auto",
            "overflowY": "hidden",
            "padding": "0 5px",
          }
        }
      >
        <Tab
          closeTab={[Function]}
          isActive={false}
          tabClick={[Function]}
          tabName="0"
        >
          <div
            style={
              Object {
                "WebkitUserSelect": "none",
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.1)",
                "boxSizing": "border-box",
                "color": "white",
                "cursor": "pointer",
                "display": "flex",
                "fontWeight": 800,
                "height": 25,
                "margin": "0 2px",
                "padding": "0 0 0 10px",
                "transform": "skew(-15deg)",
              }
            }
          >
            <div
              onClick={[Function]}
              style={
                Object {
                  "transform": "skew(15deg)",
                }
              }
            >
              0
            </div>
            <div
              onClick={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "rgba(255, 255, 255, 0.15)",
                  "display": "flex",
                  "fontWeight": 400,
                  "height": "100%",
                  "margin": "0 0 0 10px",
                  "padding": "0 10px",
                }
              }
            >
              ×
            </div>
          </div>
        </Tab>
        <Tab
          closeTab={[Function]}
          isActive={false}
          tabClick={[Function]}
          tabName="1"
        >
          <div
            style={
              Object {
                "WebkitUserSelect": "none",
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.1)",
                "boxSizing": "border-box",
                "color": "white",
                "cursor": "pointer",
                "display": "flex",
                "fontWeight": 800,
                "height": 25,
                "margin": "0 2px",
                "padding": "0 0 0 10px",
                "transform": "skew(-15deg)",
              }
            }
          >
            <div
              onClick={[Function]}
              style={
                Object {
                  "transform": "skew(15deg)",
                }
              }
            >
              1
            </div>
            <div
              onClick={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "rgba(255, 255, 255, 0.15)",
                  "display": "flex",
                  "fontWeight": 400,
                  "height": "100%",
                  "margin": "0 0 0 10px",
                  "padding": "0 10px",
                }
              }
            >
              ×
            </div>
          </div>
        </Tab>
        <AddTab
          clickAddTab={[Function]}
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "alignItems": "center",
                "backgroundColor": "rgb(241, 196, 15)",
                "boxSizing": "border-box",
                "color": "white",
                "cursor": "pointer",
                "display": "flex",
                "fontWeight": 800,
                "height": 25,
                "margin": "0 2px",
                "padding": "10px",
                "transform": "skew(-15deg)",
              }
            }
          >
            <span
              style={
                Object {
                  "transform": "skew(15deg)",
                }
              }
            >
              +
            </span>
          </div>
        </AddTab>
      </div>
      <div
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 0.4)",
            "color": "white",
            "display": "flex",
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "rgba(255, 255, 255, 0.8)",
              "height": "35px",
              "width": "35px",
            }
          }
        >
          <img
            src="https://a.ppy.sh/1_1262304000000.jpg"
            style={
              Object {
                "maxHeight": "100%",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "padding": 5,
            }
          }
        />
      </div>
    </div>
    <ChatView
      messages={Array []}
      selfNick=""
      style={
        Object {
          "flexGrow": 1,
          "marginBottom": 10,
          "overflowX": "hidden",
          "overflowY": "auto",
          "wordBreak": "break-word",
        }
      }
      userClicked={[Function]}
    >
      <div
        style={
          Object {
            "flexGrow": 1,
            "marginBottom": 10,
            "overflowX": "hidden",
            "overflowY": "auto",
            "wordBreak": "break-word",
          }
        }
      />
    </ChatView>
    <form
      onSubmit={[Function]}
      style={
        Object {
          "display": "flex",
          "flexShrink": 0,
          "height": 40,
        }
      }
    >
      <input
        onChange={[Function]}
        placeholder="Enter a message..."
        style={
          Object {
            "padding": 10,
          }
        }
        type="text"
        value=""
      />
      <input
        style={
          Object {
            "width": 200,
          }
        }
        type="submit"
        value="Send"
      />
    </form>
  </div>
</Client>
`;

exports[`Client should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    style={
        Object {
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "overflow": "hidden",
            "width": "100%",
          }
    }
>
    <JoinModal
        isOpen={false}
        joinChannel={[Function]}
        onClose={[Function]}
    />
    <div
        style={
            Object {
                "backgroundColor": "#FE4590",
                "display": "flex",
                "flexDirection": "row",
                "flexShrink": 0,
              }
        }
    >
        <div
            style={
                Object {
                    "alignItems": "flex-end",
                    "display": "flex",
                    "flexGrow": 1,
                    "overflowX": "auto",
                    "overflowY": "hidden",
                    "padding": "0 5px",
                  }
            }
        >
            <AddTab
                clickAddTab={[Function]}
            />
        </div>
        <div
            style={
                Object {
                    "backgroundColor": "rgba(255, 255, 255, 0.4)",
                    "color": "white",
                    "display": "flex",
                  }
            }
        >
            <div
                style={
                    Object {
                        "backgroundColor": "rgba(255, 255, 255, 0.8)",
                        "height": "35px",
                        "width": "35px",
                      }
                }
            >
                <img
                    src="https://a.ppy.sh/1_1262304000000.jpg"
                    style={
                        Object {
                            "maxHeight": "100%",
                          }
                    }
                />
            </div>
            <div
                style={
                    Object {
                        "alignItems": "center",
                        "display": "flex",
                        "padding": 5,
                      }
                }
            />
        </div>
    </div>
    <ChatView
        messages={Array []}
        selfNick=""
        style={
            Object {
                "flexGrow": 1,
                "marginBottom": 10,
                "overflowX": "hidden",
                "overflowY": "auto",
                "wordBreak": "break-word",
              }
        }
        userClicked={[Function]}
    />
    <form
        onSubmit={[Function]}
        style={
            Object {
                "display": "flex",
                "flexShrink": 0,
                "height": 40,
              }
        }
    >
        <input
            onChange={[Function]}
            placeholder="Enter a message..."
            style={
                Object {
                    "padding": 10,
                  }
            }
            type="text"
            value=""
        />
        <input
            style={
                Object {
                    "width": 200,
                  }
            }
            type="submit"
            value="Send"
        />
    </form>
</div>,
  "nodes": Array [
    <div
      style={
            Object {
                  "display": "flex",
                  "flexDirection": "column",
                  "height": "100%",
                  "overflow": "hidden",
                  "width": "100%",
                }
      }
>
      <JoinModal
            isOpen={false}
            joinChannel={[Function]}
            onClose={[Function]}
      />
      <div
            style={
                  Object {
                        "backgroundColor": "#FE4590",
                        "display": "flex",
                        "flexDirection": "row",
                        "flexShrink": 0,
                      }
            }
      >
            <div
                  style={
                        Object {
                              "alignItems": "flex-end",
                              "display": "flex",
                              "flexGrow": 1,
                              "overflowX": "auto",
                              "overflowY": "hidden",
                              "padding": "0 5px",
                            }
                  }
            >
                  <AddTab
                        clickAddTab={[Function]}
                  />
            </div>
            <div
                  style={
                        Object {
                              "backgroundColor": "rgba(255, 255, 255, 0.4)",
                              "color": "white",
                              "display": "flex",
                            }
                  }
            >
                  <div
                        style={
                              Object {
                                    "backgroundColor": "rgba(255, 255, 255, 0.8)",
                                    "height": "35px",
                                    "width": "35px",
                                  }
                        }
                  >
                        <img
                              src="https://a.ppy.sh/1_1262304000000.jpg"
                              style={
                                    Object {
                                          "maxHeight": "100%",
                                        }
                              }
                        />
                  </div>
                  <div
                        style={
                              Object {
                                    "alignItems": "center",
                                    "display": "flex",
                                    "padding": 5,
                                  }
                        }
                  />
            </div>
      </div>
      <ChatView
            messages={Array []}
            selfNick=""
            style={
                  Object {
                        "flexGrow": 1,
                        "marginBottom": 10,
                        "overflowX": "hidden",
                        "overflowY": "auto",
                        "wordBreak": "break-word",
                      }
            }
            userClicked={[Function]}
      />
      <form
            onSubmit={[Function]}
            style={
                  Object {
                        "display": "flex",
                        "flexShrink": 0,
                        "height": 40,
                      }
            }
      >
            <input
                  onChange={[Function]}
                  placeholder="Enter a message..."
                  style={
                        Object {
                              "padding": 10,
                            }
                  }
                  type="text"
                  value=""
            />
            <input
                  style={
                        Object {
                              "width": 200,
                            }
                  }
                  type="submit"
                  value="Send"
            />
      </form>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Client
        channelLength={5}
        channels={Array []}
        closeChannel={[Function]}
        currentChannel=""
        joinChannel={[Function]}
        makeCurrentChannel={[Function]}
        messages={Array []}
        nick=""
        openExternal={[Function]}
        sendMessage={[Function]}
        userId={1}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Client {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "displayMessages": Array [],
        "displayMessagesAmount": 50,
        "handleChangeMsg": [Function],
        "handleSubmit": [Function],
        "hideJoinModal": [Function],
        "props": Object {
          "channelLength": 5,
          "channels": Array [],
          "closeChannel": [Function],
          "currentChannel": "",
          "joinChannel": [Function],
          "makeCurrentChannel": [Function],
          "messages": Array [],
          "nick": "",
          "openExternal": [Function],
          "sendMessage": [Function],
          "userId": 1,
        },
        "refs": Object {},
        "showJoinModal": [Function],
        "state": Object {
          "msg": "",
          "showJoinModal": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "userClicked": [Function],
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          style={
                    Object {
                              "display": "flex",
                              "flexDirection": "column",
                              "height": "100%",
                              "overflow": "hidden",
                              "width": "100%",
                            }
          }
>
          <JoinModal
                    isOpen={false}
                    joinChannel={[Function]}
                    onClose={[Function]}
          />
          <div
                    style={
                              Object {
                                        "backgroundColor": "#FE4590",
                                        "display": "flex",
                                        "flexDirection": "row",
                                        "flexShrink": 0,
                                      }
                    }
          >
                    <div
                              style={
                                        Object {
                                                  "alignItems": "flex-end",
                                                  "display": "flex",
                                                  "flexGrow": 1,
                                                  "overflowX": "auto",
                                                  "overflowY": "hidden",
                                                  "padding": "0 5px",
                                                }
                              }
                    >
                              <AddTab
                                        clickAddTab={[Function]}
                              />
                    </div>
                    <div
                              style={
                                        Object {
                                                  "backgroundColor": "rgba(255, 255, 255, 0.4)",
                                                  "color": "white",
                                                  "display": "flex",
                                                }
                              }
                    >
                              <div
                                        style={
                                                  Object {
                                                            "backgroundColor": "rgba(255, 255, 255, 0.8)",
                                                            "height": "35px",
                                                            "width": "35px",
                                                          }
                                        }
                              >
                                        <img
                                                  src="https://a.ppy.sh/1_1262304000000.jpg"
                                                  style={
                                                            Object {
                                                                      "maxHeight": "100%",
                                                                    }
                                                  }
                                        />
                              </div>
                              <div
                                        style={
                                                  Object {
                                                            "alignItems": "center",
                                                            "display": "flex",
                                                            "padding": 5,
                                                          }
                                        }
                              />
                    </div>
          </div>
          <ChatView
                    messages={Array []}
                    selfNick=""
                    style={
                              Object {
                                        "flexGrow": 1,
                                        "marginBottom": 10,
                                        "overflowX": "hidden",
                                        "overflowY": "auto",
                                        "wordBreak": "break-word",
                                      }
                    }
                    userClicked={[Function]}
          />
          <form
                    onSubmit={[Function]}
                    style={
                              Object {
                                        "display": "flex",
                                        "flexShrink": 0,
                                        "height": 40,
                                      }
                    }
          >
                    <input
                              onChange={[Function]}
                              placeholder="Enter a message..."
                              style={
                                        Object {
                                                  "padding": 10,
                                                }
                              }
                              type="text"
                              value=""
                    />
                    <input
                              style={
                                        Object {
                                                  "width": 200,
                                                }
                              }
                              type="submit"
                              value="Send"
                    />
          </form>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          style={
                    Object {
                              "display": "flex",
                              "flexDirection": "column",
                              "height": "100%",
                              "overflow": "hidden",
                              "width": "100%",
                            }
          }
>
          <JoinModal
                    isOpen={false}
                    joinChannel={[Function]}
                    onClose={[Function]}
          />
          <div
                    style={
                              Object {
                                        "backgroundColor": "#FE4590",
                                        "display": "flex",
                                        "flexDirection": "row",
                                        "flexShrink": 0,
                                      }
                    }
          >
                    <div
                              style={
                                        Object {
                                                  "alignItems": "flex-end",
                                                  "display": "flex",
                                                  "flexGrow": 1,
                                                  "overflowX": "auto",
                                                  "overflowY": "hidden",
                                                  "padding": "0 5px",
                                                }
                              }
                    >
                              <AddTab
                                        clickAddTab={[Function]}
                              />
                    </div>
                    <div
                              style={
                                        Object {
                                                  "backgroundColor": "rgba(255, 255, 255, 0.4)",
                                                  "color": "white",
                                                  "display": "flex",
                                                }
                              }
                    >
                              <div
                                        style={
                                                  Object {
                                                            "backgroundColor": "rgba(255, 255, 255, 0.8)",
                                                            "height": "35px",
                                                            "width": "35px",
                                                          }
                                        }
                              >
                                        <img
                                                  src="https://a.ppy.sh/1_1262304000000.jpg"
                                                  style={
                                                            Object {
                                                                      "maxHeight": "100%",
                                                                    }
                                                  }
                                        />
                              </div>
                              <div
                                        style={
                                                  Object {
                                                            "alignItems": "center",
                                                            "display": "flex",
                                                            "padding": 5,
                                                          }
                                        }
                              />
                    </div>
          </div>
          <ChatView
                    messages={Array []}
                    selfNick=""
                    style={
                              Object {
                                        "flexGrow": 1,
                                        "marginBottom": 10,
                                        "overflowX": "hidden",
                                        "overflowY": "auto",
                                        "wordBreak": "break-word",
                                      }
                    }
                    userClicked={[Function]}
          />
          <form
                    onSubmit={[Function]}
                    style={
                              Object {
                                        "display": "flex",
                                        "flexShrink": 0,
                                        "height": 40,
                                      }
                    }
          >
                    <input
                              onChange={[Function]}
                              placeholder="Enter a message..."
                              style={
                                        Object {
                                                  "padding": 10,
                                                }
                              }
                              type="text"
                              value=""
                    />
                    <input
                              style={
                                        Object {
                                                  "width": 200,
                                                }
                              }
                              type="submit"
                              value="Send"
                    />
          </form>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Client
    channelLength={5}
    channels={Array []}
    closeChannel={[Function]}
    currentChannel=""
    joinChannel={[Function]}
    makeCurrentChannel={[Function]}
    messages={Array []}
    nick=""
    openExternal={[Function]}
    sendMessage={[Function]}
    userId={1}
/>,
}
`;
